src/code-considerations/using-unstable-lang/const-evaluation.md

```diff
@@ -0,0 +1,100 @@
+# Const evaluation
+
+Rust's compile-time code evaluation story is still in an experimental state,
+with a large amount of unstable surface area, including areas where there is
+active design work.
+
+The standard library (as of this writing) uses some amount of hacks to work
+around language limitations for some implementations: this document provides a
+framework for evaluating the cost/benefit analysis of using such workarounds to
+enable adding new `const` functions/traits or `const`-ifying existing APIs.
+
+The overall high-level guidance is to evaluate the tradeoff between enabling
+experimentation in users of the standard library (or discovery of limitations
+to these language features) and the costs to the maintenance and readability of
+the standard library. (This is broad guidance applicable to almost any feature,
+but const generics are particularly notable in terms of their potentially wide
+scope).
+
+These are good areas to consider in such an evaluation. You should think about
+whether the given PR is adding value to outweigh its costs along these
+dimensions, and add to the PR description what your justification for the usage
+is. Consider including future plans for the feature as well: if the feature is
+experimental, does it have a clear path toward stabilization in the current
+form?
+
+## Maintainability
+
+Does making this API const require that code be hand-written which was previously automatic?
+
+Does the new code require duplicating, for example through `#[cfg(bootstrap)]`,
+increasing the risk of diverging between the two implementations (and
+increasing work when bumping the bootstrap compiler)?
+
+Will updates/bugfixes to the code require further workarounds to CTFE
+limitations?
+
+Are the changes making a decision for many parts of the standard library?  For
+example, if a trait is made const, that likely means all implementations in the
+standard library have to be const too. It may not be the case that this is easy
+to accomplish (particularly if their current implementations are complex and
+need const hacks to work around limitations in CTFE).
+
+## Stability risks
+
+When making an internal (or external) API `const`, it becomes harder to
+evaluate whether a given standard library function is safe to stabilize as
+const. One of its dependencies may take a long time to stabilize for
+const-evaluation, even though its API doesn't obviously reference unstable
+surface area. 
+
+For example, a function like `[T]::get_many_mut` may rely on
+sorting to guarantee distinct indices. If sorting is made unstably const, it is
+difficult to know when looking at `get_many_mut` whether it is OK to
+const-stabilize it: you have to inspect the full dependency tree with knowledge
+of the "fully determined" const surface area.
```

Comment:
we have logic to automate this. You cannot stabilize a const fn's constness if it calls unstable const fns unless you explicitly add `#[rustc_allow_const_fn_unstable(foo, bar)]`

Reply:
Are all transitively called const fns required to be either stable or unstable by annotation? It was my impression that private helpers are sometimes const but not tagged with a const stability marker. Maybe bad memory on my part though.

Reply:
untagged functions are const stable by default, so you can't call other unstable functions unless you tag it

Reply:
hmm... I should add tests, but I'm fairly certain you can only avoid the stability marker if you're already unstable.

Reply:
This compiles fine today (just threw this into library/core/src/lib.rs). To me that looks like it's exposing in a const-stable function functionality that _might_ depend on const-eval-select (obviously this example doesn't but it's no guarantee).

Even if we forbid this, we'll want some language here about the thought process for adding const_stable to internal functions -- I suspect we're not very good at having rules there yet.

```rust
  const fn foo() {
      unsafe { crate::intrinsics::const_eval_select((), comp, run) }
  }

  const fn comp() {}
  fn run() {}

  #[stable(feature = "unreachable", since = "1.27.0")]
  #[rustc_const_stable(feature = "const_unreachable_unchecked", since = "1.57.0")]
  pub const fn hidden_danger() {
      foo();
  }
```

Reply:
ah yes. I remember that there was some issue with not being able to add those attributes to functions that don't already have `stable`/`unstable` attributes

Reply:
Cc https://github.com/rust-lang/rust/pull/103193 - looks like a practical example of this case, presumably there's more hiding.